docs/design/sum_types.md

```diff
@@ -0,0 +1,197 @@
+# Sum types
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Overview](#overview)
+-   [`choice` declarations](#choice-declarations)
+-   [User-defined sum types](#user-defined-sum-types)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
```

Comment:
Maybe it'd be helpful to draw a comparison to `enum`s? This document currently only covers the more complex cases of `choice`, not the `enum`-equivalent  scenario.

Reply:
I'm ambivalent about that. It could be a useful illustration, but it also opens up a can of worms: `choice` isn't a complete substitute for C++ `enum`, because a substantial fraction of `enum` use cases rely on the ability to convert to/from integer types, and/or the fact that an `enum` can have a value other than one of the enumerators. So it could be helpful from a teaching point of view, but it could also be counterproductive from a specification point of view.